CIME/build_scripts/buildlib.kokkos

```diff
@@ -48,62 +50,98 @@ OR
 ###############################################################################
 def buildlib(bldroot, installpath, case):
     ###############################################################################
+    # If we are using Albany/Trilinos, we must use the Kokkos that comes with it
+    if case.get_value("USE_ALBANY") or case.get_value("USE_TRILINOS"):
+        print("case is using Trilinos and will use its Kokkos")
+        return
+
+    installed_kokkos_dir = os.environ.get("Kokkos_ROOT")
+    if installed_kokkos_dir is not None:
+        # We are trying to use a pre-installed kokkos. Look for the relevant folders/libs,
+        # and if all looks good, return. Otherwise, crap out
+        kokkos_root = os.path.abspath(installed_kokkos_dir)
+        include_dir = os.path.join(kokkos_root, "include")
+
+        expect(
+            os.path.isdir(kokkos_root),
+            f"Non-existent kokkos install dir '{kokkos_root}'",
+        )
+        expect(
+            os.path.isdir(include_dir),
+            f"Missing include subfolder in kokkos install dir '{kokkos_root}'",
+        )
+        expect(
+            os.path.isfile(os.path.join(include_dir, "Kokkos_Core.hpp")),
+            f"Missing kokkos headers in '{include_dir}'",
+        )
+
+        # TODO: should I check for libs too? The problem is that the lib subfolder is
+        # often arch dependent (e.g., $prefix/lib or $prefix/lib64)... The best thing
+        # would be to run a small cmake script, that calls find_package. But E3SM's
+        # cmake build system will do that soon enough, so any error will be caught there.
+        print(f"Using pre-installed Kokkos_ROOT: {kokkos_root}")
+        return
+    else:
+        print("no value found in env for Kokkos_ROOT. building from scratch")
+
     srcroot = case.get_value("SRCROOT")
-    kokkos_dir = os.path.join(srcroot, "externals", "kokkos")
-    expect(os.path.isdir(kokkos_dir), "Missing kokkos submodule")
+    ekat_dir = os.path.join(srcroot, "libraries", "ekat")
```

Comment:
This seems to be *very* e3sm-specific. I don't think other CIME customers may need/want/have ekat (and should not be forced to clone ekat just to get kokkos). Also, ekat is in `externals/ekat` in e3sm.

Reply:
Maybe we could do something like
```py
kokkos_dir = None
kokkos_subdirs = {'core', 'container', 'simd', 'algorithm'}
for dirpath, dirnames, filenames in os.walk(os.path.join(srcroot,"externals")):
        if os.path.basename(dirpath) == 'kokkos' and 'LICENSE' in filenames and kokkos_subdirs.issubset(set(dirnames)):
		kokkos_dir = dirpath
```
This does not rely on a particular location of the kokkos folder (other than being in `ROOT/externals` somewhere). Note: checking `kokkos` dir content is necessary, as ekat has other `kokkos` subdirs.

Reply:
Thanks @bartgol . This change is actually CAM-specific because there is an ongoing project to bring the theta-l and theta-l_kokkos dycores into CAM, which requires the Kokkos library. Note that CAM does not have `$SRCROOT/externals` but `$SRCROOT/libraries` instead, where we typically save the external libraries.

I thought I needed both ekat and kokkos repo to build Kokkos library and get theta-l/theta-l_kokkos dycores compiled. But if ekat is not needed, I can remove it here to reduce complexity (also not cloned in CAM as well).

Reply:
You guys should feel free to use EKAT if you want, but it may make more sense to just add a Kokkos submodule in `srcroot/libraries/kokkos` instead of putting EKAT there. If you guys plan to actually use EKAT features, then I withdraw this comment.

Reply:
I agree with Jim. Also, you guys may want to keep up with kokkos releases at a different pace than us (faster or slower). So it is a better approach to just have `github.com/kokoks/kokkos` as a submodule (e3sm uses a fork of kokkos, following a policy of using in-house forks for tpls).

Reply:
Re: `externals` vs `libraries`. Should CIME offer this as an XML model config setting? Something like `external_tpls_src_root`?

Reply:
One thing you should consider, especially if you are going to assume CESM things like the location of the kokkos submodule, you can have this script live directly in CESM like we do for E3SM.

Reply:
Thanks @jgfouca . I looked at the `pm-cpu` machine file for EAMxx (https://github.com/E3SM-Project/E3SM/blob/master/components/eamxx/cmake/machine-files/pm-cpu.cmake#L1-L5) and it includes EKAT there. Thus I assume that we need EKAT in order to build theta-l and theta-l_kokkos dycores in CAM?

Reply:
No, the dycore does not _need_ ekat. Right now, the cmake build system relies on some ekat's cmake script to look for and possibly build kokkos. But since CAM does not use CMake, you can do away with ekat altogether. You should be able to just use the kokkos repo.

Reply:
Thanks @jgfouca and @bartgol . After talking with @jedwards4b , we decide to revoke the kokkos changes here and move this script to live in a CAM-specific place instead as you all suggest.

For @bartgol , so do I understand correctly that EKAT is needed in order to build Kokkos? And this is the reason that kokkos stays insides EKAT in the E3SM structure like `externals/ekat/extern/kokkos`?

Reply:
No, you don't need EKAT to build kokkos. Ekat provides some convenience cmake scripts (which may be convenient for us, just b/c we wrote them with US in mind, they may not be great for everyone).

The reason kokkos sits in ekat in e3sm is that ekat is its own pkg, which other projects used, so it made sense for ekat to ship it. At that point, e3sm had 2 kokkos submodules (which were not necessarily in sync): `externals/kokkos` and `externals/ekat/extern/kokkos`. We decided to rm the former, so we only had to keep track of one...

But, again, you don't need ekat.

Reply:
I should mention that the SE kokkos dycore has some duplicated code with ekat, so _at some point_ we may ekat-ify homme, so that we are more uniform across eamxx (e.g., hommexx has its own "Vector" structure, which is a simd-like struct very similar to the ekat::Pack one). But this should not affect CESM.

Reply:
Ah, thanks @bartgol for your clarification. That is clear to me now.

Maybe this is a dumb question: what is EKAT? And what is its relationship with Kokkos?

Reply:
EKAT started as a bunch of utilities _inside_ eamxx (which used to be just called "scream"). When other projects decided they could use some of those utilities, we extracted them in a pkg, which we called E3SM Kokkos Application Toolkit (EKAT), since at the time it contained mostly simd utilities and stuff related to kokkos (e.g., a tridiag solver, kokkos policies utilities, kokkos views manipulation, etc).

It then eveloved into a bunch of utilities, not just kokkos related. E.g., cmake utilities for unit testing generation, interface to a logging library, yaml parser/writer, string manipulation, etc. We never bothered changing the name, even though it should just be "E3SM C++ utils miscellanea" :)

Reply:
Gotcha! That is really good to know. Thanks for explaining it.